src/signal.cpp

```diff
@@ -0,0 +1,345 @@
+            // We can not use stop here as we do not want to raise signal.
+            should_run = false;
+            break;
+        default:
+            BOOST_ASSERT(false);
+        }
+    });
+}
+
+void
+handler_t::run() {
+    siginfo_t info;
+    should_run = true;
+    while(should_run || pending()) {
+        const int signal_num = wait_wrapper(&sig_set, &info);
+        SIG_DBG_LOG("caught signal " << signal_num);
+        if(signal_num == -1) {
+            throw std::system_error(std::error_code(errno, std::system_category()), "sigwaitinfo failed");
+        }
+        // Do not process interruption signal in case stop() was received.
+        // As we use SIGCONT there is no harm of doing this
+        if(signal_num == interrupt_signal && !should_run) {
+            continue;
+        }
+        // We need a storage to move all callbacks at one hop to prevent deadlock while executing handlers which reset themselves.
+        detailed_callback_storage tmp_storage;
+        std::lock_guard<std::mutex> guard(process_lock);
+        registered_signal_callbacks.apply([&](std::vector<detailed_callback_storage>& callbacks) {
+            if(callbacks.size() > static_cast<size_t>(signal_num)) {
+                // Move callbacks at one hop and process them outside lock
+                tmp_storage = std::move(callbacks[signal_num]);
+            }
+
+        });
+        if(tmp_storage.empty()) {
+            SIG_DBG_LOG("runnning default action ");
+            default_action(signal_num);
+        }
+        for(auto& cb_pair : tmp_storage) {
+            SIG_DBG_LOG("runnning handler");
+            cb_pair.second(std::error_code(), signal_num, &info);
+        }
+    }
+}
+
+cancellation_t
+handler_t::async_wait(int signum, detailed_callback_type handler) {
+    cancellation_t token;
+    registered_signal_callbacks.apply([&](std::vector<detailed_callback_storage>& callbacks) {
+        if(callbacks.size() <= static_cast<size_t>(signum)) {
+            callbacks.resize(signum + 1);
+        }
+        auto index = current_index++;
+        callbacks[signum][index] = handler;
+        token = cancellation_t([=]() -> bool {
+            return registered_signal_callbacks.apply([&](std::vector<detailed_callback_storage>& callbacks) -> bool {
+                auto cb_it = callbacks[signum].find(index);
+                bool result = true;
+                if(cb_it != callbacks[signum].end()) {
+                    cb_it->second(std::make_error_code(std::errc::operation_canceled), 0, nullptr);
+                    callbacks[signum].erase(cb_it);
+                } else {
+                    result = false;
+                    // In this specific situation
+                    // Wait until current signal is processed(if there is one) to be sure
+                    // That we are not processing this callback right now.
+                    std::lock_guard<std::mutex> guard(process_lock);
```

Comment:
We want to cancel handler, but it has already been scheduled. In this situation it's safer to wait until it completes in order to avoid dangerous situations in RAII cases with dangling references.